instant-clip-tokenizer-py/src/lib.rs

```diff
@@ -0,0 +1,135 @@
+    /// model.
+    #[new]
+    fn new() -> Self {
+        Tokenizer {
+            inner: instant_clip_tokenizer::Tokenizer::new(),
+        }
+    }
+
+    /// Create a new `Tokenizer` by reading the vocabulary data from the given
+    /// file name.
+    ///
+    /// The data must be in the format used by the original CLIP tokenizer
+    /// implementation from OpenAI.
+    #[staticmethod]
+    fn load(file_name: &str, max_vocabulary_size: u16) -> PyResult<Self> {
+        let f = BufReader::new(File::open(file_name)?);
+        let tokenizer = instant_clip_tokenizer::Tokenizer::with_vocabulary(f, max_vocabulary_size)?;
+        Ok(Self { inner: tokenizer })
+    }
+
+    /// Tokenize one or multiple input strings.
+    ///
+    /// Each given input string is encoded using the `encode` method and the
+    /// numeric representation written to a row in the resulting two-dimensional
+    /// numpy array of shape `(len(texts), context_length)`, with the special
+    /// `<start_of_text>` token prepended, and `<end_of_text>` appended to each
+    /// text.
+    ///
+    /// The individual input strings are lowercased before being tokenized, but
+    /// otherwise no pre-processing is performed.
+    ///
+    /// `context_length` is the maximum number of tokens per each text and
+    /// defaults to `77` which is the correct value for all current CLIP models.
+    /// If tokenization results in less than `context_length` tokens the
+    /// resulting row will be padded with trailing zeros. If tokenizing an input
+    /// text results in too many tokens, the token sequence will be truncated to
+    /// fit within the resulting row of length `context_length`, always
+    /// including the `<start_of_text>` and `<end_of_text>` marker tokens.
+    ///
+    /// The resulting array can be passed directly to the CLIP neural network.
+    fn tokenize_batch<'py>(
+        &self,
+        py: Python<'py>,
+        input: TokenizeBatchInput,
+        context_length: Option<usize>,
+    ) -> PyResult<&'py PyArray2<u16>> {
+        let context_length = context_length.unwrap_or(77);
+        if context_length < 3 {
+            return Err(PyValueError::new_err("context_length is less than 3"));
+        }
+        let result = match input {
+            TokenizeBatchInput::Single(text) => self.inner.tokenize_batch([text], context_length),
+            TokenizeBatchInput::Multiple(texts) => self.inner.tokenize_batch(texts, context_length),
+        };
+        Ok(result.into_pyarray(py))
+    }
+
+    /// Encode a `text` input as a sequence of tokens.
+    ///
+    /// The encoded token sequence does not include the special
+    /// `<start_of_text>` and `<end_of_text>` marker tokens. When these are
+    /// needed you can either use the `tokenize_batch` method instead, or add
+    /// them manually by using the `start_of_text` and `end_of_text`
+    /// methods.
+    fn encode(&self, text: &str) -> Vec<u16> {
```

Comment:
Why not yield `Token`s?

Reply:
We could return a Python iterator instead of the `Vec<u16>` - that saves us one allocation but would mean more complicated code:

`encode` would have to return a special type now:
```rust
    fn encode(&self, text: &str) -> TokenIter {
        let mut tokens = Vec::with_capacity(text.len());
        self.inner.encode(text, &mut tokens);
        TokenIter {
            iter: tokens.into_iter(),
        }
    }
```

That type, `TokenIter`, has to implement Python's iterator protocol:
```rust
#[pyclass]
struct TokenIter {
    iter: std::vec::IntoIter<instant_clip_tokenizer::Token>,
}

#[pymethods]
impl TokenIter {
    fn __iter__(slf: PyRef<Self>) -> PyRef<Self> {
        slf
    }

    fn __next__(mut slf: PyRefMut<Self>) -> Option<u16> {
        slf.iter.next().map(instant_clip_tokenizer::Token::to_u16)
    }
}
```

Do you think saving one allocation is worth the increase in complexity? Is there another way of "yielding" from Rust code that I don't know?

Reply:
My question was actually more about `u16` vs `Token` -- if we should expose a `Token` type to Python and use it here.

But the question about iterator vs `Vec` is fair, too. Maybe do a quick benchmark on this?

I'd also say the `Vec::with_capacity()` call allocating as if there will be a token per byte is not great, maybe pick something more like `len() / 3` or some such?

Reply:
> My question was actually more about u16 vs Token -- if we should expose a Token type to Python and use it here.

I would prefer to stick with `u16`. The `Token` type doesn't offer any additional functionality but is there just for type-safety in Rust. In Rust, I think it's worth having even though using the Tokenizer is a bit more work (users have to call `to_u16` to get the actual token value) because Rust-users tend to be familiar with this newtype-pattern.
In Python, however, I don't think we gain much in terms of "type-safety" but lose in API surface size and ergonomics.

> But the question about iterator vs Vec is fair, too. Maybe do a quick benchmark on this?

As expected, just calling the method is a tiny bit faster with the Iterator, but accessing all elements (I uses `sum` below for this purpose) is quite a bit slower with the Iterator-based version! Probably because all these calls to `__next__` have to cross the FFI border, I'm not sure.
In summary, we should keep the `Vec` based version :)

Benchmark:

```
In [15]: %timeit tokenizer.encode("A person riding a motorcycle.")
1.77 µs ± 10.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [16]: %timeit tokenizer.encode_iter("A person riding a motorcycle.")
1.64 µs ± 12.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [17]: %timeit sum(tokenizer.encode("A person riding a motorcycle."))
1.88 µs ± 16.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [18]: %timeit sum(tokenizer.encode_iter("A person riding a motorcycle."))
2.35 µs ± 17.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

> I'd also say the Vec::with_capacity() call allocating as if there will be a token per byte is not great, maybe pick something more like len() / 3 or some such?

Indeed, we should just use `Vec::new`, as the `instant_clip_tokenizer::Tokenizer::encode` method already has a `reserve` call and knows more about the capacity it requires.
However, this `reserve` call also allocates for the worst-case (`out.reserve(text.as_bytes().len());`). The worst-case is hit whenever we call `encode` with just a single word, so quite frequently. For sentences containing multiple words we are over-allocating here.
I would either keep that `reserve` as is, or remove it completely. I'm not a huge fan of things like `len() / 3` because it's very hard to justify why we're dividing by `3` here, why not use `len() / 4` etc.